### PR TITLE
Add reader implementation in rocksmq

### PR DESCRIPTION
**Pull request by fishpenguin**
Signed-off-by: fishpenguin <kun.yu@zilliz.com>

/kind enhancement

issue :#11976

### CODE REVIEW

File: internal/util/rocksmq/server/rocksmq/rocksmq_impl.go [hunk 1]
```diff
@@ -882,3 +885,159 @@ func (rmq *rocksmq) updateAckedInfo(topicName, groupName string, ids []UniqueID)
 	}
 	return nil
 }
+
+func (rmq *rocksmq) ReaderSeek(topicName string, msgID UniqueID) {
+	rmq.readerOffset.Store(topicName, msgID)
```
Comment (congqixia) [inline internal/util/rocksmq/server/rocksmq/rocksmq_impl.go:890]: Will there be a case that multiple reader seeks on same topic?

File: internal/util/rocksmq/server/rocksmq/rocksmq_impl.go [hunk 2]
```diff
@@ -882,3 +885,159 @@ func (rmq *rocksmq) updateAckedInfo(topicName, groupName string, ids []UniqueID)
...
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("Stop get next reader message!")
+			return ConsumerMessage{}, nil
+		default:
+			if iter.Seek([]byte(dataKey)); !iter.Valid() {
+				continue
+			}
```
Comment (congqixia) [inline internal/util/rocksmq/server/rocksmq/rocksmq_impl.go:932]: If the `dataKey`  is invalid, the block will be running forever

Comment (fishpenguin) [reply internal/util/rocksmq/server/rocksmq/rocksmq_impl.go:932]: This is by design, when invalid, Next() should block until context done.

File: internal/util/rocksmq/server/rocksmq/rocksmq_impl.go [hunk 3]
```diff
@@ -882,3 +885,159 @@ func (rmq *rocksmq) updateAckedInfo(topicName, groupName string, ids []UniqueID)
...
+	if id, ok := rmq.readerOffset.Load(topicName); ok {
+		msgID = id.(UniqueID)
+	} else {
+		return ConsumerMessage{}, fmt.Errorf("topic name = %s not exist in readerOffset", topicName)
+	}
+	readOpts.SetPrefixSameAsStart(true)
+	iter := rmq.store.NewIterator(readOpts)
+	defer iter.Close()
+	dataKey := path.Join(fixChanName, strconv.FormatInt(msgID, 10))
+
```
Comment (congqixia) [inline internal/util/rocksmq/server/rocksmq/rocksmq_impl.go:923]: Create an iterator every `Next` call is inefficient. Is it possible to combine a living `Reader` to an `Iterator` until the Reader is closed or seeked?

File: internal/util/rocksmq/server/rocksmq/rocksmq_impl.go (hunk 2 above)
Comment (congqixia) [reply internal/util/rocksmq/server/rocksmq/rocksmq_impl.go:932]: Spinning here will consume lots of CPU and disk I/O

File: internal/util/rocksmq/server/rocksmq/rocksmq_impl.go [hunk 4]
```diff
@@ -159,6 +161,7 @@ func NewRocksMQ(name string, idAllocator allocator.GIDAllocator) (*rocksmq, erro
 		storeMu:     &sync.Mutex{},
 		consumers:   sync.Map{},
 		ackedMu:     sync.Map{},
+		readers:     sync.Map{},
```
Comment (congqixia) [inline internal/util/rocksmq/server/rocksmq/rocksmq_impl.go:164]: Reader is a map from `Topic` to `Reader`. There may be a case that multiple reader on same topic. Then the logic here is messed up.

**Pull request merged**